File: initAndInstallAll.pl
#!/usr/bin/perl
use strict;
use warnings FATAL => 'all';
use Getopt::Long;
use Pod::Usage;

my $installDir="/opt/jenkinstests";
my $user;
my $man = 0;
my $help = 0;
my $remove = 0;
GetOptions(
    'help'       => \$help,
    'man'        => \$man,
    'user=s'     => \$user,
    'installDir=s' => \$installDir,
    'remove' => \$remove
) or pod2usage(2);

pod2usage(-exitval => 0, -verbose => 2) if $man;
if (!$user)
{
    print("User argument missing","\n");
    pod2usage(1);
}
print "Running with Target directory=$installDir, user:$user", "\n";
print "Not Removeing Installation Dir: $installDir" if !$remove;
if ($remove && -e $installDir) {
    print "Removeing Installation Dir:  $installDir recursively", "\n";
    system "sudo rm -Rf  $installDir"
}
if  (!-e $installDir) {
    print "Creating Installation Dir:  $installDir", "\n";
    system "sudo mkdir $installDir";
    print "Changing Owner of  Installation Dir:  $installDir", "\n";
    my $chown = "sudo chown -R ${\trim(`id -u`)}:${\trim(`id -g`)} $installDir";
    print "Executing : $chown", "\n";
    system $chown
}
system "./initLocalRepos.sh -u $user -i $installDir";
system "./initLocalRepos.sh -u $user -i $installDir";
system "./installJenkinsFilerunner.sh -i $installDir";
system "./installApscli.sh -i $installDir";
system "./runJenkinsPipeline.sh -i $installDir -a";

sub  trim { my $s = shift; $s =~ s/^\s+|\s+$//g; return $s };

__END__

=head1 NAME

 initAndInstallAll.pl

=head1 SYNOPSIS

    initAndInstallAll.pl -u USER [ -i INSTALL_DIR ] [-r]

=head1 DESCRIPTION

    Runs all installation and initialization scripts for Jenkinspipeline Tests:
    - initLocalRepos.sh : Initialization of a Gradle User Home and Mavenlocal with USER
    - installJenkinsFilerunner.sh : Installs the Jenkins Filerunner
    - installApscli.sh : Installs the serverless apscli needed for apg Patch Pipelines
    - testJenkinsPipeline.sh : Tests the Installations


=head1 ARGUMENTS

    -h|help          display this help and exit
    -u|user=USER     userid for the gitrepo, from which apg-gradle-properties will be clone. Mandatory option
    -i|installDir=INSTALL_DIR Installation Directory , defaults to "/opt/jenkinstests"
    -r|remove        removes the INSTALL_DIR before installation, defaults to false

=head1 AUTHOR

 chhex

=head1 CREDITS

=head1 TESTED

=head1 BUGS

 None that I know of.

=head1 TODO


=head1 UPDATES


=cut

File: initAndInstallAll.rb
#!/usr/bin/env ruby
# encoding: utf-8
require "slop"

opts = Slop.parse do |o|
  o.string '-u', '--username', '--user', 'Apg git user'
  o.separator ''
  o.separator 'other options:'
  o.string '-i', '--installDir', 'Jenkinsrunner Installation Dir', default: '/opt/jenkinstests'
  o.bool '-r', '--remove', 'Installation will be removed first, default false', default: false
  o.bool '-s', '--skipLocalRepo', 'Skip Initialization of Local Repos ', default: false
  o.bool '-x', '--skipFileRunner', 'Skip Installation of Jenkins File Runner ', default: false
  o.bool '-n', '--skipApsCli', 'Skip Installation of Aps serverless Cli ', default: false
  o.on '-h', '--help' do
    puts o
    exit
  end
end

if !opts[:user]
  puts "Username must be supplied"
  puts opts
  exit
end

if opts[:remove] and (opts[:skipLocalRepo] || opts[:skipFileRunner] ||  opts[:skipApsCli])
  puts "Removing installation directory  #{opts[:installDir]} contradicts skip options  "
  puts opts
  exit
end

puts "Running with User: #{opts[:user]}, remove: #{opts[:remove]}, installation Directory: #{opts[:installDir]}"

if opts[:remove] && Dir.exist?(opts[:installDir])
  puts "Deleting Installation Dir:  #{opts[:installDir]} recursively with sudo"
  system("sudo rm -Rf  #{opts[:installDir]}")
  puts "done."
end

if !Dir.exist?(opts[:installDir])
  puts "Creating Installation Dir:  #{opts[:installDir]}"
  system("sudo mkdir #{opts[:installDir]}")
  puts "Changing Owner of  Installation Dir: #{opts[:installDir]} to current user: #{opts[:user]}"
  userId = `id -u`.strip
  userGrpId = `id -g`.strip
  chownCmd = "sudo chown -R #{userId}:#{userGrpId} #{opts[:installDir]}"
  puts "Executing : #{chownCmd}", "\n";
  system(chownCmd)
end
if !opts[:skipLocalRepo]
  system("./initLocalRepos.sh -u #{opts[:user]} -i #{opts[:installDir]}")
end
if !opts[:skipFileRunner]
  system("./installJenkinsFilerunner.sh -i #{opts[:installDir]}")
end
if !opts[:skipApsCli]
  system("./installApscli.sh -i #{opts[:installDir]}")
end
system("./runTestPipeline.pl -i #{opts[:installDir]} -t runTestLibHelloWorld")
